src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -246,31 +279,276 @@ inline DocumentStatisticsAnnotationData::DocumentStatisticsAnnotationData(Freque
+    ) {
 }
 
-inline Details::DocumentStatisticsTrainingOnlyPolicy::DocumentStatisticsTrainingOnlyPolicy(StringDecorator decorator) :
-    _stringDecoratorFunc(
-        [&decorator](void) -> StringDecorator & {
-            if(!decorator)
-                throw std::invalid_argument("decorator");
+// ----------------------------------------------------------------------
+// |
+// |  Anonymous namespace
+// |
+// ----------------------------------------------------------------------
+namespace {
+
+struct IterRangeComp {
+    bool operator()(const std::tuple<std::string::const_iterator, std::string::const_iterator>& a,
+                    const std::tuple<std::string::const_iterator, std::string::const_iterator>& b) const {
+
+        std::string::const_iterator s1 = std::get<0>(a);
+        std::string::const_iterator e1 = std::get<1>(a);
+        std::string::const_iterator s2 = std::get<0>(b);
+        std::string::const_iterator e2 = std::get<1>(b);
+
+        if(std::distance(s1, e1) < std::distance(s2, e2)) {
+            return true;
+        } else if (std::distance(s2, e2) < std::distance(s1, e1)) {
+            return false;
+        }
+
+        while (s1 != e1) {
+            assert(s2 != e2);
+            if (*s2 < *s1) {
+                return false;
+            } else if (*s1 < *s2) {
+                return true;
+            }
+            ++s1;
+            ++s2;
+        }
+        return false;
+    }
+};
+
+inline Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap PruneTermFreqMap(Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap termFrequency,
+                                                                                    std::float_t minDf,
+                                                                                    std::float_t maxDf,
+                                                                                    std::float_t totalNumDocumentsFloat,
+                                                                                    nonstd::optional<std::uint32_t> topKTerms) {
 
-            return decorator;
+    using FrequencyMap = Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap;
+
+    //trim by minDf and maxDf
+    if (minDf > 0.0f || maxDf < 1.0f) {
+        FrequencyMap::iterator termFrequencyIter = termFrequency.begin();
+        while (termFrequencyIter != termFrequency.end()) {
+            std::float_t const freq = termFrequencyIter->second / totalNumDocumentsFloat;
+
+            if (freq < minDf || freq > maxDf) {
+                termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            } else {
+                ++termFrequencyIter;
+            }
+        }
+    }
+
+    //trim by topKTerms
+    if (!topKTerms.has_value() || termFrequency.empty() || *topKTerms >= termFrequency.size())
+        return termFrequency;
+
+    std::uint32_t const & topKTermsValue(*topKTerms);
+
+    std::tuple<std::uint32_t, long long> targetPack(
```

Comment:
This code could use some comments. I think that you are calculating the minimum value for the number of occurrences that should be kept and also the total number of items that should be kept?

Reply:
commented

Reply:
Eg: assuming i have an vector [2,3,3,7,7,6,5,4,4,4,4,4] and the aim is to find top k values(if k = 6), after partial sorting we know the 6th largest value is 4(the targetValue), and there are one 4 ahead of it(remainingTargetValue)